package/base-files/files/lib/functions/caldata.sh

```diff
@@ -70,7 +70,7 @@ caldata_extract_reverse() {
 	local caldata
 
 	mtd=$(find_mtd_chardev "$part")
-	reversed=$(hexdump -v -s $offset -n $count -e '/1 "%02x "' $mtd)
+	reversed=$(hexdump -v -s $offset -n $count -e '1/1 "%02x "' $mtd)
```

Comment:
This change breaks caldata at least on zyxel,nbg7815 I tested this without using any of the auxiliary functions like caldata_patch etc. purely using caldata_extract

Reply:
What is the output for this two commands?
```
hexdump -v -s 0x1000 -n 10 -e '1/1 "%02x "' /dev/mtd17
hexdump -v -s 0x1000 -n 10 -e '/1 "%02x "' /dev/mtd17
```

Reply:
I can confirm that it breaks sth. for nbg7815. I've tried it yesterday with 6.6 (vanilla, just pulling your 3 commits). I didn't look into it further. So I cannot confirm its this specific line. The output for both commands on my device is identical:

`01 00 04 04 00 00 00 00 00 80`

I just observed that within dmesg there was a line complaining "firmware not found". I'll try to make a new build in a few hours and try to collect more information if possible.

Reply:
Run this bash script and compare checksum:
```
#!/bin/sh

caldata=""

reversed=$(hexdump -v -s 0x1000 -n 131072 -e '1/1 "%02x "' /dev/mtd17)

for byte in $reversed; do
	caldata="\x${byte}${caldata}"
done
printf "%b" "$caldata" > /tmp/pre-caldata.bin

caldata=""

reversed=$(hexdump -v -s 0x1000 -n 131072 -e '/1 "%02x "' /dev/mtd17)

for byte in $reversed; do
	caldata="\x${byte}${caldata}"
done
printf "%b" "$caldata" > /tmp/pre-caldata_2.bin
```

Reply:
Sth. is wrong in this for-loop. It get stuck forever. $reversed is filled.

Reply:
I now added the 1 again and this time there was no issue, not sure what's going on. But caldata extraction on the armor g5 via hotplug when blockd is installed behaves erroneous at least.

Reply:
As a correction, this function is not used for ipq807x (it is only used for a few fritz devices).
The function is correct, but in the case of large amounts of data, the execution time is very long (in this case, over 4 minutes).
A change in the code does not affect the data.

Reply:
Sorry. I didn't have time to compile yesterday. Today I'm facing issues with gettext. I hope I can figure out the issue.

If I understand you right this is a runtime issue. I didn't wait 4 minutes for this script to finish.

But in general it would make sense and this is not related to your patch. We have already a runtime issue regarding firmware extract with blockd involved for this device (which I didn't use for my test). I will try to make a build as soon as possible and report back if I can reproduce anything. But I don't think its related to your patch after ricsc has indicated it is working. 

The first command is faster (like 2 seconds). The second takes like 8-10 seconds. So I would stick to the first.
In my experience cat is a bit faster but I don't know if cat is allowed in those POSIX/sh scripts.

Reply:
I had a issue once that `pre-caldata` was not extracted running initramfs image on other device. But didn't check what was the issue.

Reply:
Hmm. I never thought about that this could be an issue regarding testing. Anyway. I've made a clean build (tools/toolchain new) with your three commits only. Without any other modification and without luci now. I've loaded it three times through initramfs. Everything got loaded.
After that test I've flashed it to the device. Several reboots. No issue. So I guess sth. fkd things up last time or it was indeed initramfs related.

A side question: How do this fix should look for this device? I've tried:

`zyxel,nbg7815)
	caldata_extract "0:art" 0x1000 0x20000
	label_mac=$(get_mac_label)
	ath11k_patch_mac $(macaddr_add $label_mac 3) 1
	ath11k_patch_mac $(macaddr_add $label_mac 2) 2
	ath11k_patch_mac $(macaddr_add $label_mac 4) 3`

But that did nothing in regard fixing the MAC. But the device came up working with its usual random MAC's anyway.

Reply:
You must additionally use the `ath11k_fix_caldata` function. `pre-caldata` does not contain real MAC addresses for this device.

Reply:
@pwned-pixel Can you check if these commands give you correct MAC addresses?
```
hexdump -s 14 -n 6 -C /lib/firmware/ath11k/IPQ8074/hw2.0/cal-ahb-c000000.wifi.bin
hexdump -s 20 -n 6 -C /lib/firmware/ath11k/IPQ8074/hw2.0/cal-ahb-c000000.wifi.bin
hexdump -s 26 -n 6 -C /lib/firmware/ath11k/IPQ8074/hw2.0/cal-ahb-c000000.wifi.bin
```

Reply:
```
root@NBG7815:~# hexdump -s 14 -n 6 -C /lib/firmware/ath11k/IPQ8074/hw2.0/cal-ahb-c000000.wifi.bin
0000000e  00 03 7f 12 34 56                                 |....4V|
00000014
root@NBG7815:~# hexdump -s 20 -n 6 -C /lib/firmware/ath11k/IPQ8074/hw2.0/cal-ahb-c000000.wifi.bin
00000014  00 03 7f 12 34 57                                 |....4W|
0000001a
root@NBG7815:~# hexdump -s 26 -n 6 -C /lib/firmware/ath11k/IPQ8074/hw2.0/cal-ahb-c000000.wifi.bin
0000001a  00 03 7f 12 34 58                                 |....4X|
00000020
```

The last 4 digits of each are alternated on each reboot. While I do not understand why we would need label_mac here. Because label MAC is ZyXEL DEV MAC (starting with BC:CF:4F:xx:xx:xx, attached on WAN/LAN). While 00 03 7f is Atheros DEV MAC.

Reply:
MAC ids used in function starts from `1` not `0`. That's why third MAC is not set.

Reply:
Y, that what I thought already. I've tried several combinations. But all the same. Then I realized this file is meant to not change across reboots. Oh dear.
Made a "normal" build now with all my settings and applications. Everything is working now. Thank you very much for your endurance in guiding me.

Final settings I've use are:

```
	zyxel,nbg7815)
		caldata_extract "0:art" 0x1000 0x20000
		label_mac=$(get_mac_label)
		ath11k_patch_mac $(macaddr_add $label_mac 2) 1 
		ath11k_patch_mac $(macaddr_add $label_mac 3) 2 
		ath11k_patch_mac $(macaddr_add $label_mac 4) 3 
		ath11k_fix_caldata
		;;
```

I have one question left. As I've made a pull request for the MAC fix with the "ieee80211/11_fix_wifi_mac" (https://github.com/openwrt/openwrt/pull/15025) approach. Should I close it or keep it open and modify it accordingly after your patch gets accepted? What is the advise/usual procedure?

Reply:
As the interfaces are numbered from 0, I changed the parameters for function from 0 to 5.